src/ResultWriter/EnergyOutput.cpp

```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
-                                                       godunovData[i],
-                                                       drOutput[i].slip);
+        totalFrictionalWork += drOutput[i].frictionalEnergy;
+        staticFrictionalWork += computeStaticWork(timeDerivativePlus[i],
+                                                  timeDerivativeMinus[i],
+                                                  faceInformation[i],
+                                                  godunovData[i],
+                                                  drOutput[i].slip);
       }
     }
   }
-  if (usePlasticity) {
-    plasticMomentLocal = computePlasticMoment(i_meshReader, i_ltsTree, i_lts, i_ltsLut);
+  if (isPlasticityEnabled) {
+    plasticMoment = computePlasticMoment();
   }
-  int rank;
-  double totalFrictionalWorkGlobal = 0.0;
-  double staticFrictionalWorkGlobal = 0.0;
-  double plasticMomentGlobal = 0.0;
-#ifdef USE_MPI
-  MPI_Comm_rank(MPI::mpi.comm(), &rank);
-  MPI_Reduce(&totalFrictionalWorkLocal,
-             &totalFrictionalWorkGlobal,
-             1,
-             MPI_DOUBLE,
-             MPI_SUM,
-             0,
-             MPI::mpi.comm());
-  MPI_Reduce(&staticFrictionalWorkLocal,
-             &staticFrictionalWorkGlobal,
-             1,
-             MPI_DOUBLE,
-             MPI_SUM,
-             0,
-             MPI::mpi.comm());
-  MPI_Reduce(&plasticMomentLocal, &plasticMomentGlobal, 1, MPI_DOUBLE, MPI_SUM, 0, MPI::mpi.comm());
+}
+
+void EnergyOutput::computeEnergies() {
+  energiesStorage.energies.fill(0.0);
+
+  auto& totalGravitationalEnergyLocal = energiesStorage.gravitationalEnergy();
+  auto& totalAcousticEnergyLocal = energiesStorage.acousticEnergy();
+  auto& totalAcousticKineticEnergyLocal = energiesStorage.acousticKineticEnergy();
+  auto& totalElasticEnergyLocal = energiesStorage.elasticEnergy();
+  auto& totalElasticKineticEnergyLocal = energiesStorage.elasticKineticEnergy();
+
+  std::vector<Element> const& elements = meshReader->getElements();
+  std::vector<Vertex> const& vertices = meshReader->getVertices();
+
+  const auto g = SeisSol::main.getGravitationSetup().acceleration;
+
+  // Note: Default(none) is not possible, clang requires data sharing attribute for g, gcc forbids it
+#ifdef _OPENMP
+#pragma omp parallel for schedule(static) reduction(+ : totalGravitationalEnergyLocal, totalAcousticEnergyLocal, totalAcousticKineticEnergyLocal, totalElasticEnergyLocal, totalElasticKineticEnergyLocal) shared(elements, vertices, lts, ltsLut, global)
+#endif
+  for (std::size_t elementId = 0; elementId < elements.size(); ++elementId) {
+#ifdef USE_ELASTIC
+    real volume = MeshTools::volume(elements[elementId], vertices);
+    CellMaterialData& material = ltsLut->lookup(lts->material, elementId);
+    auto& cellInformation = ltsLut->lookup(lts->cellInformation, elementId);
+    auto& faceDisplacements = ltsLut->lookup(lts->faceDisplacements, elementId);
```

Comment:
so faceDisplacements are stored in a face-aligned referential, and we want to rotate it to the global cartesian referential?